plugins/out_websocket/websocket.c

```diff
@@ -0,0 +1,342 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019      The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_output.h>
+#include <fluent-bit/flb_utils.h>
+#include <fluent-bit/flb_network.h>
+#include <fluent-bit/flb_time.h>
+#include <fluent-bit/flb_upstream.h>
+#include <fluent-bit/flb_sha512.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/flb_sds.h>
+#include <fluent-bit/flb_http_client.h>
+#include <msgpack.h>
+
+#include "websocket.h"
+#include "websocket_conf.h"
+struct flb_output_plugin out_websocket_plugin;
+
+#define SECURED_BY "Fluent Bit"
+
+
+static int flb_ws_handshake(struct flb_upstream_conn *u_conn,
+                                    struct flb_out_ws *ctx)
+{
+    int ret;
+    size_t bytes_sent;
+    struct flb_http_client *c;
+
+    if (!u_conn) {
+        flb_error("[output_ws] upstream connection error");
+        return -1;
+    }
+	
+    /* Compose HTTP Client request */
+    c = flb_http_client(u_conn, FLB_HTTP_GET, ctx->uri,
+                        NULL, 0, NULL, 0, NULL, 0);
+    flb_http_buffer_size(c, ctx->buffer_size);
+    flb_http_add_header(c, "Upgrade", 7, "websocket", 9);
+    flb_http_add_header(c, "Connection", 10, "Upgrade", 7);
+    flb_http_add_header(c, "Sec-WebSocket-Key", 17, "dGhlIHNhbXBsZSBub25jZQ==", 24);
+    flb_http_add_header(c, "Sec-WebSocket-Version", 21, "13", 2);
+	
+    /* Perform request*/
+    ret = flb_http_do(c, &bytes_sent);
+	
+    if (ret != 0 || c->resp.status != 101) {
+        if (c->resp.payload_size > 0) {
+            flb_debug("[output_ws] Websocket Server Response\n%s",
+                c->resp.payload);
+        }
+        flb_http_client_destroy(c);
+        flb_upstream_conn_release(u_conn);
+        flb_debug("[out_ws] Http Get Operation ret = %i, http resp = %i", ret, c->resp.status);
+        return -1;
+    }
+    flb_http_client_destroy(c);
+    return 0;
+}
+
+static void flb_ws_mask(char *data, int len, char *mask)
+{
+    int i;
+    for (i=0;i<len;++i) {
+        *(data+i) ^= *(mask+(i%4));
+    }
+}
+
+/*
```

Comment:
is this a pending function to be implemented ?

Reply:
This function is to deal with ping/pong frame in websocket.  I will delete them for current version. 
I originally planned to use websocket ping/pong frame to maintain websocket connections since tcp connection would be dropped like 35 seconds in idle state. But it is not easily to implement within the flush function for output plugin. 
As a result, i have changed the strategy, that is to reset the state of websocket client and send handshake again, once tcp connection has been dropped.
I would add the illustration of the logic into fluent-bit doc as well.